src/approaches/grammar_search_invention_approach.py

```diff
@@ -802,22 +810,38 @@ class _HeuristicCountBasedScoreFunction(_HeuristicBasedScoreFunction):  # pylint
     the transition is optimal and the sequence is not a demo, that's
     assumed to be bad; if the transition is not optimal and the sequence
     is a demo, that's also assumed to be bad.
+
+    Also: for each successor that is one step off the atoms_sequence, if the
+    state is optimal, then check if the state is "suspicious", meaning that it
+    does not "match" any state in the demo data. The definition of match is
+    currently based on utils.unify(). It may be that this definition is too
+    strong, so we could consider others. The idea is to try to distinguish
+    states that are actually impossible (suspicious) from ones that are simply
```

Comment:
this whole thing is quite suspicious lol. there isn't really either direction of an iff between "state is suspicious per our definition" and "state is bad for optimally achieving the goal"